pkg/controllermanager/controller/exposureclass/reconciler_test.go

```diff
@@ -0,0 +1,127 @@
+	. "github.com/onsi/gomega"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/tools/record"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	fakeclient "sigs.k8s.io/controller-runtime/pkg/client/fake"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	gardencorev1alpha1 "github.com/gardener/gardener/pkg/apis/core/v1alpha1"
+	gardencorev1beta1 "github.com/gardener/gardener/pkg/apis/core/v1beta1"
+	"github.com/gardener/gardener/pkg/client/kubernetes"
+	. "github.com/gardener/gardener/pkg/utils/test/matchers"
+)
+
+var _ = Describe("Controller", func() {
+	const finalizerName = "gardener"
+
+	var (
+		ctx        = context.TODO()
+		fakeClient client.Client
+		reconciler reconcile.Reconciler
+
+		exposureClassName string
+		exposureClass     *gardencorev1alpha1.ExposureClass
+		shoot             *gardencorev1beta1.Shoot
+	)
+
+	BeforeEach(func() {
+		fakeClient = fakeclient.NewClientBuilder().WithScheme(kubernetes.GardenScheme).Build()
+
+		exposureClassName = "test-exposureclass"
+		reconciler = &Reconciler{Client: fakeClient, Recorder: &record.FakeRecorder{}}
+		exposureClass = &gardencorev1alpha1.ExposureClass{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: exposureClassName,
+			},
+		}
+
+		shoot = &gardencorev1beta1.Shoot{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      "test-shoot",
+				Namespace: "test-namespace",
+			},
+			Spec: gardencorev1beta1.ShootSpec{
+				ExposureClassName: &exposureClassName,
+			},
+		}
+	})
+
+	It("should return nil beacuse object not found", func() {
+		Expect(fakeClient.Get(ctx, client.ObjectKeyFromObject(exposureClass), &gardencorev1alpha1.ExposureClass{})).To(BeNotFoundError())
+
+		result, err := reconciler.Reconcile(ctx, reconcile.Request{NamespacedName: types.NamespacedName{Name: exposureClassName}})
+		Expect(result).To(Equal(reconcile.Result{}))
+		Expect(err).NotTo(HaveOccurred())
+	})
+
+	Context("when deletion timestamp is not set", func() {
+		BeforeEach(func() {
+			Expect(fakeClient.Create(ctx, exposureClass)).To(Succeed())
+		})
+
+		It("should ensure the finalizer", func() {
+			result, err := reconciler.Reconcile(ctx, reconcile.Request{NamespacedName: types.NamespacedName{Name: exposureClassName}})
+			Expect(fakeClient.Get(ctx, client.ObjectKeyFromObject(exposureClass), exposureClass))
+			Expect(result).To(Equal(reconcile.Result{}))
+			Expect(err).NotTo(HaveOccurred())
+			Expect(exposureClass.GetFinalizers()).Should(ConsistOf(finalizerName))
+		})
+	})
+
+	Context("when deletion timestamp is set", func() {
+		BeforeEach(func() {
+			now := metav1.Now()
+			exposureClass.DeletionTimestamp = &now
+			exposureClass.Finalizers = []string{finalizerName}
+
+			Expect(fakeClient.Create(ctx, exposureClass)).To(Succeed())
+		})
+
+		It("should do nothing beacuse finalizer is not present", func() {
+			Expect(fakeClient.Create(ctx, shoot)).To(Succeed())
+			patch := client.MergeFrom(exposureClass.DeepCopy())
+			exposureClass.Finalizers = nil
```

Comment:
Currently 
https://github.com/gardener/gardener/blob/c40753fb789348b5cbff9fcd9bfc8e0f533c36f9/pkg/controllermanager/controller/exposureclass/reconciler.go#L78-L80
is not covered by the unit test, because the object is immediately deleted by the fakeClient.

You could use
```
exposureClass.Finalizers = []string{"random"}
```
instead of setting it to nil.